pkg/utils/tsoutil/tso_dispatcher_test.go

```diff
@@ -0,0 +1,201 @@
+	count         uint32
+	doneCh        chan struct{}
+	err           error
+}
+
+func (m *mockRequest) getForwardedHost() string {
+	return m.forwardedHost
+}
+
+func (m *mockRequest) getClientConn() *grpc.ClientConn {
+	return m.clientConn
+}
+
+func (m *mockRequest) getCount() uint32 {
+	return m.count
+}
+
+func (m *mockRequest) process(forwardStream stream, count uint32) (tsoResp, error) {
+	if m.err != nil {
+		return nil, m.err
+	}
+	return forwardStream.process(0, count, 0, 0)
+}
+
+func (m *mockRequest) postProcess(countSum int64, _ int64, _ int64) (int64, error) {
+	close(m.doneCh)
+	return countSum, m.err
+}
+
+type tsoDispatcherTestSuite struct {
+	suite.Suite
+	dispatcher *TSODispatcher
+}
+
+func TestMain(m *testing.M) {
+	goleak.VerifyTestMain(m, testutil.LeakOptions...)
+}
+
+func TestTSODispatcherTestSuite(t *testing.T) {
+	suite.Run(t, new(tsoDispatcherTestSuite))
+}
+
+func (suite *tsoDispatcherTestSuite) SetupTest() {
+	tsoProxyHandleDuration := prometheus.NewHistogram(prometheus.HistogramOpts{
+		Name: "test_tso_proxy_handle_duration_seconds",
+		Help: "Histogram of TSO proxy handle duration",
+	})
+	tsoProxyBatchSize := prometheus.NewHistogram(prometheus.HistogramOpts{
+		Name: "test_tso_proxy_batch_size",
+		Help: "Histogram of TSO proxy batch size",
+	})
+	suite.dispatcher = NewTSODispatcher(tsoProxyHandleDuration, tsoProxyBatchSize)
+}
+
+func (suite *tsoDispatcherTestSuite) TearDownTest() {
+	suite.dispatcher.Stop()
+	runtime.GC() // Force garbage collection to clean up any lingering goroutines
+}
+
+func (suite *tsoDispatcherTestSuite) TestGoroutineLeakOnStreamError() {
+	re := suite.Require()
+
+	// Create a mock tso stream
+	mockStream := &mockStream{
+		delay: 10 * time.Millisecond,
+	}
+	protoFactory := &mockProtoFactory{
+		stream: mockStream,
+	}
+
+	ctx := context.TODO()
+	forwardedHost := "test-host"
+	var reqPendingCount atomic.Int64
+	var reqDispatchCount atomic.Int64
+	var wg sync.WaitGroup
+	// mock 11000 goroutines as tso proxy client
+	for range 11000 {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			dipatchReqDoneCh := make(chan struct{})
+			for range 10 {
+				req := &mockRequest{
+					forwardedHost: forwardedHost,
+					clientConn:    nil,
+					count:         1,
+					doneCh:        make(chan struct{}),
+				}
+
+				go func() {
+					// if there has goroutine leak, it will block here
+					suite.dispatcher.DispatchRequest(ctx, req, protoFactory)
+					dipatchReqDoneCh <- struct{}{}
+				}()
+
+				select {
+				case <-dipatchReqDoneCh:
+					reqDispatchCount.Add(1)
+				case <-time.After(20 * time.Second):
+					reqPendingCount.Add(1)
+				}
+			}
+		}()
+	}
+
+	wg.Wait()
+
+	re.LessOrEqual(reqPendingCount.Load(), int64(0), "There should be no pending requests after processing")
+
+	suite.dispatcher.Stop()
+	time.Sleep(5 * time.Second) // Give some time for goroutines to clean up
+	runtime.GC()                // Force garbage collection to clean up any lingering goroutines
```

Comment:
As far as I know, once the tests within a suite are completed, the test process will end directly. Are these two operations necessary?